### DISCUSSION TITLE
Blockers we encountered attempting to refactor to use getServerSideProps

### DISCUSSION THREAD

**Discussion by neilff** [vercel/next.js#21878] [Help]
My team and I recently attempted to refactor our application to use `getServerSideProps` but ran into several blockers along the way. We had hopes that `getServerSideProps` would be able to replace our usage of `getInitialProps` since we want to statically generate some of our pages. In the end we decided to avoid using `getServerSideProps` for the time being and stick with `getInitialProps`.

Our Next.js service handles a lot of pages. And because of the vast expanse of content we expose, we need to cache our pages since our long-tail is very long. Without the cache we would see a lot more traffic than we can handle hitting our internal APIs. We do this by using Redis to store as many pages as possible. We then use a 5-day TTL combined with a stale-while-revalidate approach within our Next.js custom server. The SWR approach allows pages to be refreshed after existing for 24hrs. So pages can live between 1-5 days in our cache.

Unfortunately we ran into a bunch of issues...

1. `getServerSideProps` requires [a messy `res.end` hack in order to force `renderToHTML`](https://gist.github.com/leanazulyoro/23c6581fb4379ec311e3bc8538715687#file-server-js-L33-L54) to return HTML. Without this, pushing cached values to Redis is not possible. I'm hoping there is more official solution here in the future. From what I have seen of Incremental Static Regeneration, I don't think it would scale well for us given that we need to have a 5-day TTL with our pages, and we are shipping 3-4x per day. My take on it is that the cache bust caused by a new deployment would cause problems for us. (I'd happy to be proven wrong here)

2. The `/_next/data` endpoint request from `getServerSideProps` is made directly against the main domain. We run multiple Next.js applications on our domain, one owns the root, and then we also have other apps that own a handful of routes off the root domain. We use an ingress to determine which service sees what requests. This is fine in a `getInitialProps` world, but with `getServerSideProps` we now have to determine which service sees which `/_next/data` requests. We could use an ingress to determine this, but it could get quite messy depending on how many services we are running. Another solution appears to be to [use a `basePath`](https://nextjs.org/docs/api-reference/next.config.js/basepath), however that doesn't work for us since we have some services which own multiple routes off our route domain. We would have to spin up a new Next.js service for each route, and unfortunately that isn't trivial for us to perform.

3. When a user client-side navigates, `getServerSideProps` is run purely on the server, while `getInitialProps` is run on the client-side. When using `getServerSideProps` Next.js uses the `/_next/data` route to request the server run the `getServerSideProps` code. We found out that this won't work with our strategy for a couple of reasons:
- Since we use a [commit hash as our build ID](https://nextjs.org/docs/api-reference/next.config.js/configuring-the-build-id) our endpoint is versioned (e.g. `/_next/data/<hash>/...`), and since we might have older clients floating around (due to being eventually consistent with our caching model) those users are going to see a failure occur when they try to request `/_next/data` of an older version of our app
- Even if we didn't use a build id, we could run into problems where version A of the client hits `/_next/data` but the server is running version B, and therefore something within the `getServerSideProps` code might have changed thus returning an incorrect response to client with version A (e.g. imagine a GraphQL property being removed from our page-level query)
Neither of these issues exist with `getInitialProps` since it is run directly from the client, it knows what data it requires from the server and reaches out to it directly.

With that being said, I'm interested to know if `getInitialProps` will be deprecated in the future? 

The [official docs hint at `getServerSideProps` being the recommended](https://nextjs.org/docs/api-reference/data-fetching/getInitialProps) way to go but unfortunately for reasons stated above we can't seem to make it work.

As well, I'd be interested to know what the vision is for scaling Next.js to serve millions of pages? I've seen Incremental Static Regeneration mentioned but that doesn't appear to scale well if you're constantly busting your cache with new deployments 🤔  

I'm all ears for any ideas on how to resolve our problems.

**Comment by neilff** [comment]
@timneutkens Wonder if you have any insight here?

**Reply by jaybytez** [reply]
To resolve our problem with basePath and i18n not working well together.  We have removed basePath and in order to do this, we will be using a more custom buildId since that buildId is in the `/_next/data` path.  So each project will have our normal version (has like you mentioned) in the buildId but also the context that we would have had in basePath.  If we had two applications like blogs and products, we would have buildIds for each like `blogs__1.1.0` and `product_1.1.0` and then use our web server reverse proxy configuration to route our `/blogs` and `/_next/data/blogs__` requests to the blogs application, while routing `/product` and `/_next/data/product__` requests to the product application.  Not sure if this helps.

**Reply by neilff** [reply]
@jaybytez Thanks for the details there! I appreciate the response. We were considering something roughly the same. 🤔 

The downside for us that we discovered was that since we have many versions of our client running out in the wild (because we have a 5-day TTL in our cache), when a client like `blogs__1.1.0` hit our service, but the service was running a newer version like `blogs__2.0.0` it wouldn't be able to fulfil the response for `blogs__1.1.0` if the newer version had an updated page props response. 

Next.js luckily does force a server-side refresh in this case, but we really didn't want to break client-side routing for outdated clients, especially when we potentially have a large chunk of our traffic behind a few versions.

We discussed running services to handle older client requests (eventually consistent even on the server), but figured that would cost us way more money than it would be worth when `getInitialProps` solves the problem OOTB.
